File: sorting my life (sort algarithoms)/merge.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

// array to be sorted
int data[] = { 5,2,6,1134,13413,54,1,6,89,23,6,8,34,73,4,327,5,324 };

// merge function
void merge(int *array){
    int size = sizeof(array)/sizeof(array[0]);
    int mid = size/2;
    int left[mid];
    int right[size-mid];
    for (size_t i = 0; i < mid; i++){
        left[i] = array[i];
    }
    for (size_t i = mid; i < size; i++){
        right[i] = array[i];
    }
    merge(left);
    merge(right);
    int i = 0;
    int j = 0;
    int k = 0;
    while(i < mid && j < size-mid){
        if (left[i] < right[j]){
            array[k] = left[i];
            i++;
        }else{
            array[k] = right[j];
            j++;
        }
        k++;
    }
    while(i < mid){
        array[k] = left[i];
        i++;
        k++;
    }
    while(j < size-mid){
        array[k] = right[j];
        j++;
        k++;
    }
    for (size_t i = 0; i < size; i++){
        printf("%d ", array[i]);
    }
}
void main(){
    merge(data);
}

File: sorting my life (sort algarithoms)/merge.js


let merge_sort = (arr) => {
    if (arr.length === 1) {
        return arr;
    }
    let middle = Math.floor(arr.length / 2);
    let left = arr.slice(0, middle);
    let right = arr.slice(middle);
    return merge(merge_sort(left), merge_sort(right));
}

let merge = (left, right) => {
    let result = [];
    let index_left = 0;
    let index_right = 0;
    while (index_left < left.length && index_right < right.length) {
        if (left[index_left] < right[index_right]) {
            result.push(left[index_left]);
            index_left++;
        }
        else {
            result.push(right[index_right]);
            index_right++;
        }
    }
    return result.concat(left.slice(index_left)).concat(right.slice(index_right));
}

let arr = [5, 4, 3, 2, 1];
console.log(merge_sort(arr));
